### ISSUE TITLE
unable to successfully deploy PVC on raspberry pi

### ISSUE THREAD

**Issue by bohnejoe** [kubernetes-sigs/nfs-subdir-external-provisioner#36]
Hi,

I am trying to deploy nextcloud on k3s and store my data on a nfs-share.

I already setup the k3s on a (for now) single raspberry pi via [k3s-ansible](https://github.com/k3s-io/k3s-ansible) which worked perfectly fine. I tried to setup a PV+PVC for a NFS share and eventually found that I should use this nfs provisioner.

I have
- setup a NFS mount (verified via showmount -e IP) and was able to mount it manually via mount.nfs and mount -t nfs
- cloned this repository
- modified deploy/helm/values.yaml (nfs.server, nfs.path, repository: quay.io/external_storage/nfs-client-provisioner-arm)
- also changed kubeVersion to ">=1.9.0 <1.21.0" in deploy/helm/Chart.yaml because k3s version is "v1.20.0+k3s2"
- deployed the helm chart in a namespace "nfs"

The deployment seems to be working fine.
When I try to deploy deploy/test-claim.yml all I get is 

> Normal  ExternalProvisioning  4m33s (x263 over 69m)  persistentvolume-controller  waiting for a volume to be created, either by external provisioner "fuseim.pri/ifs" or manually created by system administrator

```
$ kubectl describe deployment nfs-subdir-external-provisioner -n nfs
Name:               nfs-subdir-external-provisioner
Namespace:          nfs
CreationTimestamp:  Wed, 13 Jan 2021 19:31:07 +0100
Labels:             app=nfs-subdir-external-provisioner
                    app.kubernetes.io/managed-by=Helm
                    chart=nfs-subdir-external-provisioner-3.0.0
                    heritage=Helm
                    release=nfs-subdir-external-provisioner
Annotations:        deployment.kubernetes.io/revision: 1
                    meta.helm.sh/release-name: nfs-subdir-external-provisioner
                    meta.helm.sh/release-namespace: nfs
Selector:           app=nfs-subdir-external-provisioner,release=nfs-subdir-external-provisioner
Replicas:           1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:       Recreate
MinReadySeconds:    0
Pod Template:
  Labels:           app=nfs-subdir-external-provisioner
                    release=nfs-subdir-external-provisioner
  Service Account:  nfs-subdir-external-provisioner
  Containers:
   nfs-subdir-external-provisioner:
    Image:      quay.io/external_storage/nfs-client-provisioner-arm:v3.1.0-k8s1.11
    Port:       <none>
    Host Port:  <none>
    Environment:
      PROVISIONER_NAME:  cluster.local/nfs-subdir-external-provisioner
      NFS_SERVER:        192.168.0.227
      NFS_PATH:          /mnt/data
    Mounts:
      /persistentvolumes from nfs-subdir-external-provisioner-root (rw)
  Volumes:
   nfs-subdir-external-provisioner-root:
    Type:      NFS (an NFS mount that lasts the lifetime of a pod)
    Server:    192.168.0.227
    Path:      /mnt/data
    ReadOnly:  false
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   nfs-subdir-external-provisioner-598475ff9b (1/1 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  65s   deployment-controller  Scaled up replica set nfs-subdir-external-provisioner-598475ff9b to 1
```
```
$ kubectl describe pvc test-claim -n nfs
Name:          test-claim
Namespace:     nfs
StorageClass:  nfs-client
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: cluster.local/nfs-subdir-external-provisioner
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type    Reason                Age               From                         Message
  ----    ------                ----              ----                         -------
  Normal  ExternalProvisioning  7s (x3 over 12s)  persistentvolume-controller  waiting for a volume to be created, either by external provisioner "cluster.local/nfs-subdir-external-provisioner" or manually created by system administrator
```

I tried to deploy without helm but still the same issue

Thank you for your help and sorry if I am missing any information. Quite new to the k8s universe ;)

**Comment by yonatankahana** [comment]
> also changed kubeVersion to ">=1.9.0 <1.21.0" in deploy/helm/Chart.yaml because k3s version is "v1.20.0+k3s2"

the reason that we set it to `>=1.9.0 <1.20.0` is because `1.20` is not supported in the latest official image.
we have already merged fix for that to `master` (#26) and if you'll build your own image (`make container`) you should be able to make it work by setting the values.yaml to use your image.

i hope that pretty soon we will see release of newer version that will contain this fix and when it will happen we will also update the helm chart kubeVersion range accordingly.

**Comment by bohnejoe** [comment]
Thank you for the fast reply! I'll try to build the image and test again :)

**Comment by bohnejoe** [comment]
is there any ETA for a new release?

**Comment by yonatankahana** [comment]
unfortunately no... you can follow the progress in #15

**Comment by bohnejoe** [comment]
really hard to follow, what is happening. Maybe open a github project and make a column with all required issues for the first release?
as i understand the first release will not include the fix for this issue? at least i can not find anything in the 4.0.0-rc1?

**Comment by texasbobs** [comment]
Has anyone built a working image?  I tried to build one on my pi cluster but it gets an error at execution time. `standard_init_linux.go:219: exec user process caused: exec format error`
I'm sure I did something wrong. :) 
On the pi, I...
-  installed golang 1.15 and docker on one of the pis 
- cloned this repo
- ran `make build`
- ran `make container` 
- tagged and push my new image up to dock hub
- updated my yaml

**Comment by yonatankahana** [comment]
> on my pi cluster

what image you use, are you using the [`-arm` image](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/blob/master/deploy/deployment-arm.yaml)?

**Comment by texasbobs** [comment]
> > on my pi cluster
> 
> what image you use, are you using the [`-arm` image](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/blob/master/deploy/deployment-arm.yaml#L24)?

No.  I am on 1.20 of k3s so that image won't work.  I was trying to build a new one.

**Comment by kmova** [comment]
@texasbobs - can you try with this image: `gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner:canary`

The above is generated and pushed by K8s cloud builds with the latest changes. You can see more details here:
https://console.cloud.google.com/gcr/images/k8s-staging-sig-storage/GLOBAL/nfs-subdir-external-provisioner?authuser=0&gcrImageListsize=30

**Comment by texasbobs** [comment]
> @texasbobs - can you try with this image: `gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner:canary`
> 
I'm not sure that was made for ARM7.  I tried it though and it reports back with `Failed to pull image "gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner:canary": rpc error: code = NotFound desc = failed to pull and unpack image "gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner:canary": no match for platform in manifest: not found`

**Comment by yonatankahana** [comment]
@kmova arm7 is 32bit, what does it take to add it to the `BUILD_PLATFORMS` and multi-arch image?

**Comment by texasbobs** [comment]
I just checked and was wrong.  The rpi 4 uses ARM v8.

**Comment by yonatankahana** [comment]
try to use `gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner:arm64-linux-v4.0.0`

**Comment by neinkeinkaffee** [comment]
I have a similar problem setup as @bohnejoe in that I want to use the nfs storage provisioner for persistence in my k3s Raspberry Pi 4 cluster. When I deploy the provisioner on k3s with `gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner:arm64-linux-v4.0.0` I get an `exec format error`. It might be due to the fact that most OS for the Raspberry Pi 4 are still 32-bit, including Raspbian OS Buster that I am using. Do you also provide an arm7 build?

**Comment by neinkeinkaffee** [comment]
I am getting an `exec format error` when trying to use `gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner:arm64-linux-v4.0.0` with k3s on Raspberry Pi 4. It might be because most OS for Raspberry Pi 4 are still 32-bit, although the hardware is 64-bit arm8. Would be it be possible to add arm7 to the `BUILD_PLATFORMS`?

**Comment by hamid-elaosta** [comment]
> I have a similar problem setup as @bohnejoe in that I want to use the nfs storage provisioner for persistence in my k3s Raspberry Pi 4 cluster. When I deploy the provisioner on k3s with `gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner:arm64-linux-v4.0.0` I get an `exec format error`. It might be due to the fact that most OS for the Raspberry Pi 4 are still 32-bit, including Raspbian OS Buster that I am using. Would be it be possible to add arm7 linux to `BUILD_PLATFORMS`?


You are correct. The Raspbian OS userspace is still 32bit. You can run Ubuntu 20.04 aarch64 or use a 32bit ARM image.

**Comment by neinkeinkaffee** [comment]
Thanks @hamid-elaosta. I used the multi-arch build in this fork https://github.com/groundhog2k/nfs-subdir-external-provisioner to check and it worked.

**Comment by tomasz-kuzemko** [comment]
I had the same issue with not being able to run the Docker image on a Raspberry Pi 4. For anyone else stuck on this, this is how I was able to build the 32-bit ARM image in a clone of this repository:

```sh
$ make build BUILD_PLATFORMS="linux arm"
$ DOCKER_CLI_EXPERIMENTAL=enabled docker buildx build --tag nfs-subdir-external-provisioner:arm-linux-canary --platform=linux/arm --file Dockerfile --build-arg binary=./bin/nfs-subdir-external-provisioner  .
# last command is optional - import the image to target node running k3s if you don't have any image registry
$ docker save nfs-subdir-external-provisioner:arm-linux-canary | ssh PI_HOSTNAME sudo ctr i import -  
```

@kmova for a permanent solution I would also like to ask to extend `BUILD_PLATFORMS` with `linux arm`.

**Comment by bohnejoe** [comment]
/close

it works for me now with the following helm parameters:
```
image:
  repository: gcr.io/k8s-staging-sig-storage/nfs-subdir-external-provisioner
  tag: arm64-linux-canary
```

Thanks alot :)
